### ISSUE TITLE
DISCUSS: remove remaining usages of character codes in array reprs

### ISSUE THREAD

**Issue by ngoldbaum** [numpy/numpy#25508]
I realize this is a little late in the game to get this done, but in case there's appetite for it, I think it might be a good idea to change array reprs to no longer use dtype character codes in NumPy 2.0. Specifically, this happens with `np.str`, `np.bytes_`, and `np.void`:

```
>>> np.array(["hello", "world"])
array(['hello', 'world'], dtype='<U5')
```

I think it would fit in better with the numeric dtypes to instead print this:

```
array(['hello', 'world'], dtype=np.dtypes.StrDType(5))
```

And similarly for the bytes dtype and unstructured voids.

One wrinkle is we don't actually support doing this with unstructured voids right now:

```
In [2]: np.dtypes.VoidDType(10)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 np.dtypes.VoidDType(10)

TypeError: Preliminary-API: Flexible/Parametric legacy DType '<class 'numpy.dtypes.VoidDType'>' can only be instantiated using `np.dtype(...)`
```

But I think updating the reprs of `bytes_` and `str` are probably more important if we don't want to allow a similar syntax with `void`.

**Comment by mhvk** [comment]
The disadvantage of your very long dtype is that it distracts greatly from the actual content of the array, which surely should be the focus. This in contrast to what is the case for numerical values, where one gets `dtype=int8`, etc. I guess removing "np.dtypes." goes some way towards shortening it, but having `dtype=...DType` still feels a bit cluttered and distracting.

Trying to get inspiration for something better from numerical cases, I noticed the situation is currently quite a mess in that even for those the character-representation gets shown in all but the default circumstance:
```
np.arange(4, dtype='i4')
# array([0, 1, 2, 3], dtype=int32)

np.arange(4, dtype='>i4')
# array([0, 1, 2, 3], dtype='>i4')

np.arange(4, dtype='i4').view('i4,i4')
# array([(0, 1), (2, 3)], dtype=[('f0', '<i4'), ('f1', '<i4')])

np.arange(4, dtype=np.dtype(np.longdouble).newbyteorder())
# array([0., 1., 2., 3.], dtype='>f16')
```
The last example also shows how tedious it is to get a different byte order without using the character codes (how is this improving code clarity? or can it be done more elegantly without ">f16"?) 

Anyway, I guess the move away from character + number of bytes for numerical types is incomplete, leaving us in a situation that seems worse than what we had... Though I do agree that for string/bytes it is annoying how, e.g., the 4 in `U4` is not the number of bytes... It certainly would be good to have something better...

**Comment by ngoldbaum** [comment]
Good point about the non-default numerical dtypes.

> The last example also shows how tedious it is to get a different byte order without using the character codes (how is this improving code clarity? or can it be done more elegantly without ">f16"?)

It's certainly not as concise as `<F16`, but one could imagine allowing something like `np.dtypes.Float16DType(byteorder="le")` (just as an example API). And similarly allow any other parameters to show up in the dtype and array repr using keyword-style
syntax.

Maybe there another possible API we could come up with that's less cryptic than character codes and modifier symbols and numbers but a little less wordy than instantiating a camel-cases class name?

**Comment by mhvk** [comment]
Indeed, I looked at trying to instantiate a byte-order using the new `DType` classes, but couldn't. In fact, they don't seem to be documented beyond a list at https://numpy.org/doc/stable/reference/routines.other.html

Let me ping @seberg to see what the "right" way to do this would be...

**Comment by seberg** [comment]
Adding it to the class constructor is the right thing, I just didn't do it because I was defensive when first introducing the classes at all.  (Mainly for the sake of legacy user dtypes, and also to not discuss API; at the time it didn't matter anyway.)

But, as much as I dislike the string shorthands since I don't like user dtypes feeling compelled to use strings.  I can't disagree that the full repr is long.

Now... IMO, non-native byte order is a not an issue:  It's rare, so to me it seems fine if the `repr` is a mouthful!  Avoiding the `i4` notation is nice because it most users shouldn't have to know that `int32` and `i4` are the same thing.
(OK, they always have to learn something new, but I guess I like the idea of exposing users to the the idea that dtypes are class instances.)

Unfortunately, that doesn't solve the conundrum.  For anything except legacy strings I would be happy to use the class constructor _except_ where we already use the `int32` short-hand in the array repr, the array repr doesn't 100% preserve dtypes anyway.

For strings where every string is parametric I am not sure if `np.dtypes.StrDType(5)` is nice.  We could try to shorten the class names: `np.StrDType(5)` (allowing just `StrDTypes(5)` in the repr) or `np.dtypes.Str(5)`?  But not sure that is great either.

---

The last thought I have is going back to string land and accepting mini-languages (which I dislike, but...) by allowing `str[5]`, `str[5, >]` and maybe `int32[>]` to match the `m8[ns]` notation.  I.e. a generic dtype string + parameters notation.
TBH, I don't really love these mini string languages, so if the class constructors can be made a bit less of a mouthfull that might be nice?

**Comment by rgommers** [comment]
> I guess removing "np.dtypes." goes some way towards shortening it,

That's standard I'd say, namespaces are rarely included in reprs. In NumPy, we have `array`, `masked_array`, `finfo`, etc. - all no namespaces, also not if the object itself isn't in the main namespace.

> Anyway, I guess the move away from character + number of bytes for numerical types is incomplete, leaving us in a situation that seems worse than what we had

I believe it's already better, and I'm +1 on Nathan's idea here. 

> Now... IMO, non-native byte order is a not an issue: It's rare, so to me it seems fine if the repr is a mouthful! 

Agreed. Character codes are bad, and we should avoid them as much as possible. Doing that for all the basic "array with normal dtype" cases is an improvement. Dealing with corner cases like non-native endianness (which is really unimportant, and rarely used - other array libraries don't even support it at all, it just doesn't come up normally) would be nice, but should not stop us from improving the reprs for normal usage.

Re all the other shortening ideas: none of them look great. I don't see any issue with:
```
array(['hello', 'world'], dtype=StrDType(5))
```

**Comment by ngoldbaum** [comment]
I think the next step is to take a look at the C dtype formatting code for legacy numeric dtypes, see what is actually being printed, and identify cases where we're still emitting character codes. If it's impossible to create a dtype using the dtype classes with the same parameters, we should add those parameters to the dtype classes.

As part of #25507 I made it so the dtype classes get autogenerated documentation pages and will show up in the docs search, so that should also improve discoverability if we're dropping the full import path.

Despite what I said yesterday in the issue description about shipping this in numpy 2.0, I don't think this needs to be a numpy 2.0 thing so it doesn't need to happen immediately, we've changed reprs in the past already, and we have facilities in the formatting machinery to preserve a legacy repr for doctests.

**Comment by mhvk** [comment]
> (which is really unimportant, and rarely used - other array libraries don't even support it at all, it just doesn't come up normally

I don't think this is correct. Any data format has to either choose endianness or ensure it can deal with both. In astronomy, essentially all data is stored as FITS, which is big-endian, so astronomers at least encounter it frequently (even if the classes that deal with accessing FITS data mostly hide this). A quick search shows that astronomy is not alone: netCDF is also bigendian.

> > Anyway, I guess the move away from character + number of bytes for numerical types is incomplete, leaving us in a situation that seems worse than what we had
> 
> I believe it's already better

In what way? Even ignoring endianness, any user who encounters a structured dtype already has to understand two different representations of he same thing (`int32` and `i4`), and here we are suggesting to add a third, related to what is in `np.dtypes`. I remain confused about the dismissal of the mini-language - it is happily related to formatting and the alternative just is very long-winded. Also, indeally a `repr` gives something that the class can be instantiated with. This is now true for structured arrays but not for regular ones.

```
In [2]: np.arange(4, dtype='i4').view('i4,i4')
Out[2]: array([(0, 1), (2, 3)], dtype=[('f0', '<i4'), ('f1', '<i4')])

In [3]: np.array([(0, 1), (2, 3)], dtype=[('f0', '<i4'), ('f1', '<i4')])
Out[3]: array([(0, 1), (2, 3)], dtype=[('f0', '<i4'), ('f1', '<i4')])

In [4]: np.arange(4, dtype='i4')
Out[4]: array([0, 1, 2, 3], dtype=int32)

In [5]: np.array([0, 1, 2, 3], dtype=int32)
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[5], line 1
----> 1 np.array([0, 1, 2, 3], dtype=int32)

NameError: name 'int32' is not defined
```

Anyway, it is a path chosen and that's fine - let's just try to make things consistent (and concise, ideally).

**Comment by mhvk** [comment]
Aside: don't get me wrong on the new dtype organization, they are great (and were needed; just wish I had some real time to explore the unit aspect further!). I also like the idea to add arguments to allow different endianness on instantiation (though "be" may be worse than ">" as it is (even) less descriptive and doesn't easily allow for "native").

And @ngoldbaum's path forward seems sensible, of having a careful look at the formatting code.

Also another suggestion on the names: since the dtype classes all live in `np.dtypes`, can we just remove `DType` from their class names? After all, "Namespaces are one honking great idea -- let's do more of those!"

**Comment by rgommers** [comment]
It seems like you have a different mental model from me about what's important for repr's @mhvk. The main criteria I'd apply are:

1. Reprs are easy to read and understand
2. Reprs should use idiomatic code constructs, because users learn from them
3. Reprs should roundtrip (this is more a constraint that applies to any choice)

The above are imho more important than internal consistency between, for example, regular single dtypes and structured dtypes. That's nice to have too, but secondary.

For the above, it seems clear to me that users do not understand character codes (even people with numpy/scipy commit bits get confused by them regularly), hence using them breaks (1) in a major way. And it's bad for (2) as well, since their use is very much non-idiomatic. Hence, any place where we get rid of them is a win.

>  (even if the classes that deal with accessing FITS data mostly hide this).

Exactly - it's only a few data formats where this comes up, and even then it's mostly simply handled transparently. I can't think of a single case of having had to do explicit byte-swapping in ~20 years of numpy usage. And I think that's true for the vast majority of numpy users (not sure if that's 98% or 99.9%, but it'll be in that direction). 

> In what way? Even ignoring endianness, any user who encounters a structured dtype already has to understand two different representations of he same thing `(int32` and `i4`)

Same answer - quite niche, and secondary to me to improvements for the code that regular users read and write on a daily basis. The main use case of structured dtypes has long been superseded by dataframe libraries, so we're down to occasional special-case usage. The average user may never encounter them.

> and here we are suggesting to add a third, related to what is in np.dtypes. 

I think the point is that for each dtype the goal is to use the most idiomatic expression of that dtype. For 32-bit integers that's `int32` (never `i4`), and for a fixed-length string it's `StrDType(5)` now (right?). This one is perhaps less important, because it's rare to explicitly select a fixed-length size, that's usually inferred. But for the new variable-length string dtype I guess it'll be common.

> Also another suggestion on the names: since the dtype classes all live in `np.dtypes`, can we just remove `DType` from their class names? After all, "Namespaces are one honking great idea -- let's do more of those!"

Seems reasonable to me - can't remember if this was discussion before though (I'm guessing yes).

**Comment by ngoldbaum** [comment]
I worry a little that `np.dtypes.Int8` is too similar to `np.int8`, and might make the confusion between scalar types and dtypes worse. That said, it does follow normal python API patterns where a camel-cased name is a type, and that might be enough to signify that it's a different thing.

**Comment by mhvk** [comment]
@ngoldbaum - indeed, a little confusing since one does not have `isinstance(np.int8, np.dtypes.Int8DType)` -- though if dtype instances gained a `__call__` method they perhaps could be :smiley_cat: It would make things more concise, though.

@rgommers - my frustration is really mostly with the current lack of consistency. And I do use structured arrays a lot...

**Comment by seberg** [comment]
@mhvk except then you have multiple scalar types due to byte-order...  You could fake the `isinstance()` call to work one way or another but I am not really convinced.

Right, I can't disagree that for structured dtypes the/a "mini language" *is* useful.

> Seems reasonable to me - can't remember if this was discussion before though (I'm guessing yes).

I honestly doubt it.  It has the downside of potentially forcing lower case for scalar types if you use the same pattern.
Dropping the namespace is fine, it worries me a bit that everyone will have to at least once google how to find `StrDType`, though (even those who use NumPy for many years who are surprised when they see it).

**Comment by seberg** [comment]
Sorry, posted early (doesn't matter) and closed by accidental touchpad use.
